src/hotspot/share/prims/jvmtiExport.cpp

```diff
@@ -1690,20 +1690,47 @@ void JvmtiExport::continuation_yield_cleanup(JavaThread* thread, jint continuati
   }
 }
 
-void JvmtiExport::post_object_free(JvmtiEnv* env, jlong tag) {
-  Thread *thread = Thread::current();
+void JvmtiExport::post_object_free(JvmtiEnv* env, GrowableArray<jlong>* objects) {
+  assert(objects != NULL, "Nothing to post");
+  assert(env->is_enabled(JVMTI_EVENT_OBJECT_FREE), "checking");
+
+  Thread* thread = Thread::current();
+  if (thread->is_VM_thread()) {
+    post_object_free_on_vm_thread(env, objects);
+  } else {
+    post_object_free_on_java_thread(env, objects);
+  }
+}
+
+void JvmtiExport::post_object_free_on_vm_thread(JvmtiEnv* env, GrowableArray<jlong>* objects) {
+  EVT_TRIG_TRACE(JVMTI_EVENT_OBJECT_FREE, ("[?] Trg Object Free triggered" ));
+  EVT_TRACE(JVMTI_EVENT_OBJECT_FREE, ("[?] Evt Object Free sent"));
 
-  if (thread->is_Java_thread() && JavaThread::cast(thread)->is_in_VTMS_transition()) {
+  jvmtiEventObjectFree callback = env->callbacks()->ObjectFree;
+  if (callback != NULL) {
+    for (int index = 0; index < objects->length(); index++) {
+      (*callback)(env->jvmti_external(), objects->at(index));
+    }
+  }
+}
+
+void JvmtiExport::post_object_free_on_java_thread(JvmtiEnv* env, GrowableArray<jlong>* objects) {
+  JavaThread *javaThread = JavaThread::current();
+
+  if (javaThread->is_in_VTMS_transition()) {
     return; // no events should be posted if thread is in a VTMS transition
   }
```

Comment:
Chris suggested to convert this if-statement to assert.
The concern is we want to be sure the debug agent does not skip any CLASS_UNLOAD events.
Though, I feel that a probability to get the assert triggered is very low.

Reply:
> Chris suggested to convert this if-statement to assert.
> The concern is we want to be sure the debug agent does not skip any CLASS_UNLOAD events.
> Though, I feel that a probability to get the assert triggered is very low.

Yes, I think it should be an assert, but was not suggesting that change as part of this PR since it is pre-existing. We can file a separate CR for it.

Reply:
Yes, I agree that it should be another CR.